### ISSUE TITLE
Paws::S3::PutBucketLifecycle Content-MD5 is required

### ISSUE THREAD

**Issue by xpix** [pplu/aws-sdk-perl#181]
Logical roblem, the user can't set an ContentMD5 value, cuz he don't know the content of the generated XML:

w/o Content-MD5:
Response: `<Message>Missing required header for this request: Content-MD5</Message>`

with Content-MD5:
Response: `<Message>The Content-MD5 you specified was invalid.</Message><Content-MD5>1</Content-MD5>`

So, i wrote an ugly patch at Paws::Net::RestXmlCaller (don't want to make a Pulltrequest, cuz my Solution is just a patch):

```
 diff -Naur /opt/gmp-perl/perls/perl-5.24.0/lib/site_perl/5.24.0/Paws/Net/RestXmlCaller.pm Paws_tmp/Net/RestXmlCaller.pm
--- /opt/gmp-perl/perls/perl-5.24.0/lib/site_perl/5.24.0/Paws/Net/RestXmlCaller.pm      2017-05-05 20:36:34.000000000 +0000
+++ Paws_tmp/Net/RestXmlCaller.pm       2017-08-01 13:06:18.517351000 +0000
@@ -199,9 +199,6 @@

     my $url = $self->_api_endpoint . $uri; #in Paws::API::EndPointResolver

-    #TODO: I'm not sure if any of the REST style APIs want things as query parameters
-    $request->parameters({ $self->_to_querycaller_params($call) });
-
     $request->url($url);
     $request->method($call->_api_method);

@@ -209,8 +206,27 @@
       $request->content($xml_body);
     }

+    my %params = $self->_to_querycaller_params($call);
+
+    #TODO: I'm not sure if any of the REST style APIs want things as query parameters
+    $request->parameters(\%params);
+
     $self->_to_header_params($request, $call);

+    # Patch to set corrrect XML MD5
+    # ------------
+    die "Patch needed for Paws: Version 0.32 this is version: " . $Paws::VERSION
+       if($Paws::VERSION ne '0.32');
+
+    if($params{ContentMD5}){
+      use MIME::Base64 qw(encode_base64); # please check also https://github.com/pplu/aws-sdk-perl/blob/b98886ade4fb78dfbdf6816f15de8a394b943566/lib/Paws/Net/S3APIRequest.pm#L63
+      use Digest::MD5 'md5';              # Can't use string ("<LifecycleConfiguration><Rule><N"...) as a SCALAR ref while "strict refs" in use at /opt/gmp-perl/perls/perl-5.24.0/lib/site_perl/5.24.0/Paws/Net/S3APIRequest.pm line 63
+      my $content_md5 = encode_base64( md5( $request->content ) );
+      chomp $content_md5;
+      $request->headers->header( "Content-MD5" => $content_md5 );
+    }
+    # ------------
+
     if ($call->can('_stream_param')) {
       my $param_name = $call->_stream_param;
       $request->content($call->$param_name);

```

**Comment by pplu** [comment]
Hi,

  Even in the form of a "patch", it helps figure out what's needed / missing, and the best thing is that it's working code.

  Thanks for your contribution 👍

**Comment by xpix** [comment]
Hi, @pplu ,

do you try to fix this issue in the next version, i need this information for our project. If you like, i can write a better looking fix :)

**Comment by sven-schubert** [comment]
solution: https://github.com/pplu/aws-sdk-perl/pull/189

**Comment by glenveegee** [comment]
FWIW: The 0.35 release has broken PutObject (see issue #210). I have a workaround in place in PR #211.

**Comment by pplu** [comment]
Closing, as we fixed this in 0.36 (shipped to CPAN some time ago)
